google/auth/credentials.py

```diff
@@ -236,8 +223,113 @@ def with_non_blocking_refresh(self):
         self._use_non_blocking_refresh = True
 
 
+class CredentialsWithTrustBoundary(Credentials):
+    """Abstract base for credentials supporting trust boundary.
+
+    A credential with trust boundary will carry a trust boundary info as a
+    cache. The cache value can be either or fetch from a global lookup or
+    injected via ``set_trust_boundary``.
+
+    Upon we apply credential headers to a request, a trust boundary value will
+    be added as a header. GFE will use the value to routing.
+    """
+
+    def apply(self, headers, token=None):
+        """Trust boundary value will be a cached value from global lookup.
+
+        The response of trust boundary will be a list of regions and a hex
+        encoded representation.
+
+        An example of global lookup response:
+
+        .. code-block::
+
+            {
+                "locations": [
+                    "us-central1", "us-east1", "europe-west1", "asia-east1"
+                ],
+                "encoded_locations": "0xA30"
+            }
+        """
+        super().apply(headers, token)
+        if self.trust_boundary is not None:
+            headers["x-allowed-locations"] = self.trust_boundary["encoded_locations"]
```

Comment:
Is it enough to store just the encoded locations? Where does locations get used?